File: LeetCodeAlgos/Python/middleOfLinkedList.py
# Given the head of a singly linked list, return the middle node of the linked list.
# If there are two middle nodes, return the second middle node.
from typing import *

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cnt, run = 0, head
        while(run):
            cnt+=1
            run = run.next
        run = head
        cnt = int(cnt/2)
        for i in range(0, cnt):
            run = run.next
        return run

s = Solution()
s.middleNode(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))) ))
s.middleNode(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, ListNode(6))))) ))

File: LeetCodeAlgos/Rust/middle-of-linked-list.rs
// Given the head of a singly linked list, return the middle node of the linked list.

// If there are two middle nodes, return the second middle node.



#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}
impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

struct Solution {}
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut cnt: i32 = 0;
        let mut run = head.clone();
        while let Some(node) = run {
            cnt += 1;
            run = node.next;
        }
        run = head;
        cnt = cnt / 2;
        for _ in 0..cnt as usize {
            run = run.unwrap().next;
        }
        run
    }
}